### ISSUE TITLE
All and None in Flags enum should be respectively 0 and bitwise OR of all individual values

### ISSUE THREAD

**Issue by antonioaversa** [SonarSource/sonar-dotnet#8862]
Originated from this [community thread](https://community.sonarsource.com/t/false-positive-for-rule-s4070-unsigned-zero-complements-not-recognized-as-all-option/109423/3).

Definitions of `All` in an enum decorated as `[Flags]` like the following one can introduced unwanted/surprising behaviors.
```cs
[Flags]
public enum X2 : ulong
{
    None = 0,
    A = 1 << 0,
    B = 1 << 1,
    All = ulong.MaxValue, // or 0xffffffffffffffff, or ((((ulong)1) << 63) - 1) << 1 | 1, ...
}
```

The main reasons are that such definition of `All`:
- breaks the property of `Flags` enums that the bitwise OR of all individual values (defined as powers of 2, i.e. values `v` for which a non-negative integer `e` exists such that `v == 1 << e`) should give the `All` value
  - that is, there is no undeclared individual value in the enum
- it breaks the [`F`](https://learn.microsoft.com/en-us/dotnet/standard/base-types/enumeration-format-strings#f-or-f) and [`G`](https://learn.microsoft.com/en-us/dotnet/standard/base-types/enumeration-format-strings#g-or-g) modifiers of the `ToString` method when the bitwise OR of all individual values is passed

As an example, consider the following two enums, declared as `[Flags]`:
```cs
[Flags]
public enum X1 : ulong
{
    None = 0,
    A = 1 << 0,
    B = 1 << 1,
    All = A | B,
}

[Flags]
public enum X2 : ulong
{
    None = 0,
    A = 1 << 0,
    B = 1 << 1,
    All = ulong.MaxValue, // 0xffffffffffffffff
}
```
Here is how `X1` and `X2` behave:
```cs
Console.WriteLine((X1.A | X1.B) == X1.All);  // True
Console.WriteLine((X1.A | X1.B).ToString()); // All
Console.WriteLine((X1.All).ToString());      // All
Console.WriteLine((X1.All).ToString("G"));   // All
Console.WriteLine((X1.All).ToString("F"));   // All

Console.WriteLine((X2.A | X2.B) == X2.All);  // False
Console.WriteLine((X2.A | X2.B).ToString()); // A, B
Console.WriteLine((X2.All).ToString());      // All
Console.WriteLine((X2.All).ToString("G"));   // All
Console.WriteLine((X2.All).ToString("F"));   // All
```
While `X1` respects the properties described above, `X2` doesn’t.
As a result `X2.A | X2.B is not X2.All`, so there are least n undeclared values `X2.C1`, `X2.C2`, …, `X2.Cn` such that `X2.A | X2.B | X2.C1 | X2.C2 | ... | X2.Cn = X2.All`.
This would create potential bugs in code, such as the following:
```cs
void AMethodReceivingX2(X2 x2)
{
    if (x2 == X2.All) { /* Do something specific */ }
}

// ...
AMethodReceivingX2(X2.A | X2.B); // I passed all declared flags, yet it's not the same as X2.All!
```
Therefore, while it may be tempting to conceptually define `All` as `~None` for convenience and as it simplifies future maintenance of the enum, it’s more correct to:
- either not define the All flag explicitly, when not needed
- or define it as the combination of all individual values, when it is needed

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1611](https://sonarsource.atlassian.net/browse/NET-1611)

[NET-1611]: https://sonarsource.atlassian.net/browse/NET-1611?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
